File: DFS_BFS/784-letterCasepermutation.cpp
/**
 * Given a string s, you can transform every letter individually to be lowercase or uppercase to create another string.

Return a list of all possible strings we could create. Return the output in any order.

 

Example 1:

Input: s = "a1b2"
Output: ["a1b2","a1B2","A1b2","A1B2"]
Example 2:

Input: s = "3z4"
Output: ["3z4","3Z4"]
 * 
*/

class Solution {
public:
    vector<string> letterCasePermutation(string s) {
        vector<string> ans;
        dfs(s,0,ans);
        return ans;
    }
    private:
        void dfs(string& S,int s,vector<string>& ans){
            if(s==S.length()){
                ans.push_back(S);
                return;
            }
            dfs(S,s+1,ans);
            if(!isalpha(S[s])) return;
            S[s]^=(1<<5);
            dfs(S,s+1,ans);
            S[s]^=(1<<5);
        }
};

File: DFS_BFS/784-letterCasepermutation.py
def letterCasepermutation(s):
    ans=[]
    n=len(s)
    def dfs(idx,S):
        if idx==n:
            ans.append(S)
            return 
        
        dfs(idx+1,S)
        if not S[idx].isalpha():
            return
        ll=list(S)
        ll[idx]=chr(ord(ll[idx])^(1<<5))
        S="".join(ll)
        dfs(idx+1,S)
        # s[idx]^=(1<<5)
        ll=list(S)
        ll[idx]=chr(ord(ll[idx])^(1<<5))
        S="".join(ll)

    dfs(0,s)
    return ans



        
s = "a1b2"
print(letterCasepermutation(s))